### ISSUE TITLE
Construction with undetermined length

### ISSUE THREAD

**Issue by RubenKelevra** [Helium314/StreetComplete#601]
I've noticed the features, where you can enter that a section of a road is under construction. 

But weirdly you have to enter an end of construction. I mean there's a construction site and a general construction sign and that's it. 

So it's impossible to answer for me when they will be finished.

Is the idea that we go to them and ask? 🤔 

![Screenshot_2024-08-15-16-18-32-315-edit_de westnordost streetcomplete expert](https://github.com/user-attachments/assets/c330204c-eb92-4a3f-9b27-eebf871f746a)

On very large construction sites, like bridges there's sometimes a information board which states "projected end of construction may 2028" or something like that. But large construction projects are never finished at the time which is posted on them.

So it's hardly helpful to have the road marked as driveable without any review by the date I've entered like 3 years ago. 

I can't see this feature being useful in any way. 

**Proposed Solution**
I just want to flag the road as under construction and use the quest which asks if the construction is completed to answer that it's still under construction.

**Comment by mnalis** [comment]
> Is the idea that we go to them and ask? 🤔

It might often (but not always) be signed... If it is not, one should perhaps better leave the note (see below on why), so some armchair mapper can look it up (in my experience, municipality/construction/news website will often have estimated/promised end dates) and update it appropriately.

> But large construction projects are never finished at the time which is posted on them.

That doesn't mean the date is useless -- it gives you information when information should be rechecked (either via armchair, or resurveyed). With provided date, people can set up [automated checks](https://wiki.openstreetmap.org/wiki/Key:opening_date#Practical_use) to remind them when it should be opened/rechecked.

> So it's hardly helpful to have the road marked as driveable without any review by the date I've entered like 3 years ago.

Ummm, that seems like **exactly the reason** why the end-of-construction date should be provided and is very useful?
Otherwise, once marked as construction, the road might be marked as unusable for many years after if was reconstructed (until some advanced OSM mapper comes over and notices the issue).

In fact, that is the reason that I usually don't even bother marking short-term reconstructions (to be completed in a month or so), but just leave a note: with the (usually monthly or so) offline downloads and chances that people won't remember to fix its status back in timely manner, just marking it as a construction is creating more trouble than it is fixing (especially when alternative routes are significantly worse)...

If anything, I'd recommend using date-based [conditional restrictions](https://wiki.openstreetmap.org/wiki/Conditional_restrictions) instead of `highway=construction`, as they will "auto-fix" themselves if people forget to remove them on time.

**Comment by Helium314** [comment]
> But weirdly you have to enter an end of construction.

You can use the tag editor. Changing to `highway=construction` is very easy to do, while tagging with end time is more prone to errors when done manually, and thus deserves a separate answer.

> So it's hardly helpful to have the road marked as driveable without any review by the date I've entered like 3 years ago.

Then you should complain to whoever is marking the road as driveable solely based on the planned finish date you tagged 3 years ago.

> I just want to flag the road as under construction

Use the tag editor. Roads already have a _lot_ of what was previously called _other answers_, and I only want to extend the list with answers are either very common, or complicated to enter manually.

I agree with @mnalis that opening a note could be useful in such a case.

> If anything, I'd recommend using date-based [conditional restrictions](https://wiki.openstreetmap.org/wiki/Conditional_restrictions) instead of `highway=construction`, as they will "auto-fix" themselves if people forget to remove them on time.

In wiki it's recommended to use `highway=construction` if construction is scheduled to take more than 6 to 9 months, and SCEE follows this (I think I set the `access:conditional` limit to 6 months).

**Comment by RubenKelevra** [comment]
@mnalis wrote

> It might often (but not always) be signed

This might be true from where you're mapping, but absolutely not the case in Germany, as explained above.

> If anything, I'd recommend using date-based [conditional restrictions](https://wiki.openstreetmap.org/wiki/Conditional_restrictions) instead of `highway=construction`, as they will "auto-fix" themselves if people forget to remove them on time.

That's a terrible practice, and I don't know who came up with that or why. It leads to road be "auto-fixed" despite nobody actually have checked that, and routers which will route through closed roads.

I can't name a single construction project which had a projected signed and was actually finished at the projected date. So it will lead to broken routing, until the next map update and many frustrated users which don't understand why their navigation system is suddenly routing them through a closed road.

Apart from that, there's the logical issue with using the tag, as a construction site being not an access restriction. Which is kind of important for first responders to differentiate between "we may drive through there" or "we can't because there's just no road or bridge to drive on".

> > But weirdly you have to enter an end of construction.
> 
> You can use the tag editor. Changing to `highway=construction` is very easy to do, while tagging with end time is more prone to errors when done manually, and thus deserves a separate answer.

Yeah that's what I did. But doing it 20 times in a row or something like this isn't ideal from a usability perspective. 


> > So it's hardly helpful to have the road marked as driveable without any review by the date I've entered like 3 years ago.
> 
> Then you should complain to whoever is marking the road as driveable solely based on the planned finish date you tagged 3 years ago.

I can't follow your train of thought. I thought if I use the option "construction until" in SCEE will exactly do that, isn't it? 

I mean what is SCEE tagging there if I say: Construction until "in 3 years", because that's what's written on the sign? 

@Helium314 wrote:
> > I just want to flag the road as under construction
> 
> Use the tag editor. Roads already have a _lot_ of what was previously called _other answers_, and I only want to extend the list with answers are either very common, or complicated to enter manually.
> 
> I agree with @mnalis that opening a note could be useful in such a case.

Sure, but in this case, there's no need for this function in the first place. A note can also contain a construction-end-date.

> > If anything, I'd recommend using date-based [conditional restrictions](https://wiki.openstreetmap.org/wiki/Conditional_restrictions) instead of `highway=construction`, as they will "auto-fix" themselves if people forget to remove them on time.
> 
> In wiki it's recommended to use `highway=construction` if construction is scheduled to take more than 6 to 9 months, and SCEE follows this (I think I set the `access:conditional` limit to 6 months).

So if I enter "seven months" it will switch to `highway=construction` instead of `access:conditional`?

---

How about renaming this to "It's under construction..." and then ask the user if there's a posted date? If the user answers yes, the date is added as `construction_end_expected=YYYY-MM-DD`, which the "is it still under construction?" quest could parse and not show up before this date.

**Comment by mcliquid** [comment]
> I can't name a single construction project which had a projected signed and was actually finished at the projected date.

This is very different. In my federal state, every national construction site is communicated with a start and end date and the end dates are always respected. That has to be the case, otherwise European transit traffic would not be guaranteed.
A current example is the [Arlbergtunnel](https://www.asfinag.at/bauen-erhalten/bauprojekte/s-16-arlberg-schnellstrasse-sanierung-tunnel-arlberg/). This was already closed to all traffic in 2023 from 26 April to 10 October. The route was reopened to traffic the next day. The tunnel is currently closed again from 15 April to 22 November 2024 and will definitely be open again on 23 November. A delay will not be tolerated.

**Comment by Helium314** [comment]
> > If anything, I'd recommend using date-based [conditional restrictions](https://wiki.openstreetmap.org/wiki/Conditional_restrictions) instead of `highway=construction`, as they will "auto-fix" themselves if people forget to remove them on time.
> 
> That's a terrible practice, and I don't know who came up with that or why. It leads to road be "auto-fixed" despite nobody actually have checked that, and routers which will route through closed roads.

The alternative would be offline routers not routing through open roads when using `highway=construction` for a 1 week construction.
Btw "routers which will route through closed roads" is a very common case because _many_ construction sites are not tagged.

> Apart from that, there's the logical issue with using the tag, as a construction site being not an access restriction.

I think people usually are forbidden from entering construction sites, so it _is_ an access restriction. Though one could argue that the value should rather be `private`.

> Yeah that's what I did. But doing it 20 times in a row or something like this isn't ideal from a usability perspective.

I just tried, and changing a highway value to `contruction` is less than 5 seconds. Juding by that, I estimate discussing this issue uses up more time than adding this option will ever save.

> > > So it's hardly helpful to have the road marked as driveable without any review by the date I've entered like 3 years ago.
> > 
> > 
> > Then you should complain to whoever is marking the road as driveable solely based on the planned finish date you tagged 3 years ago.
> 
> I can't follow your train of thought. I thought if I use the option "construction until" in SCEE will exactly do that, isn't it?

SCEE does not set the value back to whatever it was before when 3 years have passed.

> I mean what is SCEE tagging there if I say: Construction until "in 3 years", because that's what's written on the sign?

Please use the undo funciton to check what is tagged.

> Sure, but in this case, there's no need for this function in the first place. A note can also contain a construction-end-date.

Yes, of you ignore @mnalis comment "so some armchair mapper can look it up"... but then you can argue that every other answer can also be given as note.

> So if I enter "seven months" it will switch to `highway=construction` instead of `access:conditional`?

Yes. You brought the example with 3 years, so I assumed you would know what happens when you use a date 3 years from now.

> How about renaming this to "It's under construction..." and then ask the user if there's a posted date? If the user answers yes, the date is added as `construction_end_expected=YYYY-MM-DD`, which the "is it still under construction?" quest could parse and not show up before this date.

It's work for me (I really dislike working on Android UI stuff) and for translators. Just for a minimal time save.

**Comment by mnalis** [comment]
_(reasoning for wontfix explained more than 3 weeks ago, and no further comments since then, so I'll close this)_

**Comment by RubenKelevra** [comment]
> The alternative would be offline routers not routing through open roads when using `highway=construction` for a 1 week construction.

That's a non-issue, as you'll reach your destination with a slightly longer travel time. That's a much less worse case than an offline router trying to route you through a closed road for a week.

> It's work for me (I really dislike working on Android UI stuff) and for translators. Just for a minimal time save.

Is a popup really that difficult to implement? 🤔

**Comment by mnalis** [comment]
> That's a non-issue, as you'll reach your destination with a slightly longer travel time. 

I'd say it hugely depends on the road.  :man_shrugging: As an (somewhat extreme) example, if it is the road crossing border, you could easily be rerouted hundreds of kilometers  to next border crossing... Instead of just noticing on-site road traffic rerouting signs and taking a slight detour (just like you would do if changes were not mapped in OSM, or if you didn't have/use GPS map at all). So that counterexample is significantly worse with your idea...

But even if they were equally problematic in routing regard; the main disadvantage (to me) is that with your idea the damage / incorrect data could persist for years (or more), instead of any inaccuracy being ***guaranteed to auto-fix*** in a week (as it does with current SCEE solution).

> That's a much less worse case than an offline router trying to route you through a closed road for a week.

If it is common that roadworks take a lot longer than announced in your area and you feel that _"router trying to route you through a closed road for a week"_ is common and hugely problematic result of it, and rerouting via other roads is usually not problematic at all in your area; then **you should simply overestimate the completion dates** when choosing the SCEE answer that it is under construction. 

E.g. if is says it will be finished in less then a week, and it is likely to take a week and  half or even two, then you could answer that it will take 3 (or even 4) weeks just to be sure. That way you'd get best of both worlds - it will be guaranteed auto-fix itself (if nobody fixes it manually upon completion), and (due to overly extended dates) it will never try to route you through a still closed road. Win-win, eh?

But if you still strongly think that auto-fixing is bad even with extended dates in you specific case (which I would strongly recommend that you consider instead ); then you could also relatively easily change it with SCEE in just a such undated way as you want; so I'd have to agree with above comment:

> _I just tried, and changing a highway value to `construction` is less than 5 seconds. Judging by that, I estimate discussing this issue uses up more time than adding this option will ever save._

Especially as it's highly unlikely you will be marking hundreds of closed roads per day, every day (when using raw tag editor might get somewhat taxing). And hey, that 5 seconds is just for the first time, once the value gets into "recent values", it's going to be even faster than that.

For example, I think I've already invested enough time explaining in this discussion why it's not worth it to last me for at least 20 years worth of construction road tagging savings in SCEE (and Croatia has  a lot of roads under construction) :sweat_smile: 

> Is a popup really that difficult to implement? 🤔

Apparently, and depending how good you are in Kotlin. For (another extreme) example, trying to add one popup in https://github.com/streetcomplete/StreetComplete/pull/5453 is taking me months (and I still haven't figured it out; as I don't know Kotlin). But bigger problem here is that investing that work is just not a good idea...
Anyway, as the saying goes, _nothing is ever hard for man who doesn't have to do it himself_ :smiley:

**Comment by RubenKelevra** [comment]
> I'd say it hugely depends on the road. 🤷‍♂️ As an (somewhat extreme) example, if it is the road crossing border, you could easily be rerouted hundreds of kilometers to next border crossing... Instead of just noticing on-site road traffic rerouting signs and taking a slight detour (just like you would do if changes were not mapped in OSM, or if you didn't have/use GPS map at all). So that counterexample is significantly worse with your idea...

If you flip the idea of having a couple of hundred km drive in the wrong direction, just to find out that a road is closed, and having to find a better route yourself, because your routing software doesn't know about this is definitely worse than having to drive a different way because your routing software is not up to date and thus doesn't know that the road has been opened again.



> But even if they were equally problematic in routing regard; the main disadvantage (to me) is that with your idea the damage / incorrect data could persist for years (or more), instead of any inaccuracy being _**guaranteed to auto-fix**_ in a week (as it does with current SCEE solution).

There's no damages or incorrect data, if you enter a construction zone. And we got a quest to do exactly that: Review that a road isn't a construction zone.

There's no need for a robotic action to unlock roads at random times, which are mostly never met.

As I said initially, most construction zones here are pushed back further and further, which would lead to the road being opened for routing multiple times over the years, until someone finds it and closes it again, for the now newly posted ending date.

The reason for that, contrary to the example in example by @mcliquid is that the posted date is just a projection and not a hard contract at which it should be done.

So to sum it up:

- In most cases there is no date posted, meaning I have to edit dozens of ways in the tag editor to close them.
- if a date is posted it's 99% of the time inaccurate and the access tag would open it again, meaning frustrated users find themselves routed through a closed route, which was automatically opened again.
- The rare amount of cases a route is reopened at the specified date are so high value, that nobody would "miss" to reopen them, so worst case scenario would be offline routing for a week or so is off.

As far as I can tell we're using the worst possible solution right now for this problem.

**Comment by mnalis** [comment]
Anyway, since the discussion has reactivated, I'll reopen this so other people can find it; but I personally don't have much time the rehash the arguments - I think both sides of the issue have well represented arguments pro and contra.

If someone willing to create PR themselves comes along; I suggest they submit their plan in this issue before commencing work (so maintenance burden and other issues can be appraised before they invest too much time in particular implementation). 

In the meantime: happy mapping! :world_map: :heart:

**Comment by RubenKelevra** [comment]
TL;DR given the arguments presented here, I revise my first post and sum up the best solution (IMHO) as:

User hits "There is a construction":

Q: _Is this road closed?_ 

A: **No** ➡️ `construction=minor`

A: **it's complicated** ➡️ Note

A: **Yes** ⬇️

Q: _Is there a posted end date for this construction?_

A: **No** ➡️ `highway=construction` + `construction=<old highway type>` + checkdate 

A: **Yes** ⬇️
A: Date Enter Dialog ⬇️

Q: _Is there a projected end date or a fixed date which is likely to be met?_

A: **projected date** ➡️ `highway=construction` + `construction=<old highway type>` + check date + `construction_end_expected=YYYY-MM-DD`

A: **fixed date** ➡️ (and less than 6 month out) access restriction with date

**Comment by mcliquid** [comment]
If I haven't miscounted, it takes a total of four quests to get to the end - including the initial one. In my opinion, that's too many quests / questions for just one construction site.
Maybe I have the wrong picture in my head, but under `construction=minor` I just imagine that a manhole cover at the side of the road is being replaced and therefore two orange pylons are standing at the side of the road. The traffic can still move without restriction. If my picture is correct, I wonder what relevance this has for OSM. Is this intended to influence the routing?

To cut a long story short: Isn't it enough to add to the current function (“It's under construction”) for all road quests so that the current `highway` value is changed to `construction` + sub-tag? At the moment, only `access:conditional` is set here accordingly, without any `construction`-tag.

StreetComplete and SCEE will then regularly ask if the construction site is still there anyway.

And everything else can be entered directly in the tag editor in my opinion. This is less work than answering four quests per way.

**Comment by RubenKelevra** [comment]
@mcliquid yeah, that's the optimum if we wanna cut dev time for this low.

`construction=minor` is everything which isn't a complete road closure. I put it sometimes on construction zones which block half the directions and install temporary traffic lights, as routers may wanna use it to determine longer travel times for a road.

If we don't wanna put this in, the question should be "is this road completely closed due to construction (in both directions)?" - as everything below is `construction=minor`.

At least that should be asked, to confirm that the road is not closed for your mental image due to confusion by the user.

And since we already ask this yes/no question we could just put yes/no/it's complicated in there and tag construction=minor if not.

**Comment by mnalis** [comment]
> it takes a total of four quests to get to the end - including the initial one. In my opinion, that's too many quests / questions for just one construction site.

agreed, too much!

> At the moment, only access:conditional is set here accordingly, without any construction-tag.

not correct; behaviour depends on the duration of the works chosen; see https://github.com/Helium314/SCEE/issues/601#issuecomment-2295374931

**Comment by mcliquid** [comment]
> not correct; behaviour depends on the duration of the works chosen

You're right, I overlooked that.
